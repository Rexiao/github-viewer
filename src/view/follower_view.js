import 'react-native-gesture-handler'
import React from 'react'
import { Button, Text, View } from 'react-native'

/**
 * follower screen
 * @param {Object} navigation react native navigation
 */
function FollowerScreen ({ navigation }) {
  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Text>empty</Text>
      <Button onPress={() => navigation.goBack()} title="Go back profile" />
    </View>
  )
}

export { FollowerScreen }
