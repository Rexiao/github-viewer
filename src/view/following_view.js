import 'react-native-gesture-handler'
import React from 'react'
import { Button, Text, View } from 'react-native'

/**
 * following screen
 * @param {Object} navigation react native navigation
 */
function FollowingScreen ({ navigation }) {
  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Text>empty</Text>
      <Button onPress={() => navigation.goBack()} title="Go back profile" />
    </View>
  )
}

export { FollowingScreen }
