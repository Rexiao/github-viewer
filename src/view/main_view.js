import 'react-native-gesture-handler'
import React, { useEffect, useState } from 'react'
import { Image, SafeAreaView, FlatList, Button, StyleSheet, Text, View, ActivityIndicator } from 'react-native'
import token from './../models/token'
import { Profile, queryText } from './../models/profile'

/**
 * function for home screen
 * @param {Object} props props.nav react navigation
 *                       props.needFetch if the view need actually fetch data from github
 *                       props.needLoadScreen if the view only show activity indicator
 *                       props.fakeData the fake data used in view if needFetch is false
 */
function HomeScreen (props) {
  const [isLoading, setLoading] = useState(true)
  const [data, setData] = useState([])
  const variables = {}

  /**
   * item function used in renderItem
   * @param {Object} props l:key
   *                       r:value
   *                       nav: react navigation
   */
  const Item = function itemFunction (props) {
    let content
    if (props.l === 'avatarUrl') {
      content = <Image source={{ uri: props.r }} style={{ width: 400, height: 400 }}/>
    } else if (props.l === 'public_repo_count') {
      content = <Button
        title= {'public repo number: ' + props.r}
        onPress={() => props.nav.navigate('Repo')}
      />
    } else if (props.l === 'followers_count') {
      content = <Button
        title= {'followers number: ' + props.r}
        onPress={() => props.nav.navigate('follower')}
      />
    } else if (props.l === 'followings_count') {
      content = <Button
        title= {'followings number: ' + props.r}
        onPress={() => props.nav.navigate('following')}
      />
    } else {
      content = <Text style={styles.title}>{props.l} : {props.r}</Text>
    }
    return (
      <View style={styles.item}>
        {content}
      </View>
    )
  }
  /**
   * wrapper used in flat list
   * @param {Object} props l:key
   *                       r:value
   *                       nav: react navigation
   */
  const renderItem = ({ item }) => (
    <Item l={Object.keys(item)[0]} r={item[Object.keys(item)[0]]} nav={props.nav} />
  )
  useEffect(() => {
    if (props.needFetch) {
      Profile.fetchWrapper(token, queryText, variables, setData, setLoading)
    } else {
      // console.log('null becuase of fake!')
      if (props.needLoadScreen) {
        setLoading(true)
      } else if (!props.fakeData) {
        setLoading(false)
        setData(null)
      } else {
        setLoading(false)
        setData(props.fakeData)
      }
    }
  }, [])
  return (
    <SafeAreaView style={styles.container}>
      {isLoading ? <ActivityIndicator/> : (
        !data ? (<Text>load data fails. If the reason is related to network, check console.error. Otherwise, it is becuase of invalid data format</Text>) : (
          <FlatList
            data={data}
            renderItem={renderItem}
            keyExtractor={(item, index) => index.toString()}
          />
        )
      )}
    </SafeAreaView>
  )
}

/**
 * wrapper for home screen
 * @param {Object} navigation react navigation
 */
function HomeScreenWrapper ({ navigation }) {
  return (
    <HomeScreen needLoadScreen={false} nav={navigation} needFetch={true } fakeData={[
      {
        avatarUrl: 'https://avatars3.githubusercontent.com/u/42951357?u=a648f7ff3392b1139e1208bc1b6774d249b25f7b&v=4'
      },
      { name: 'Ang' },
      { login: 'Rexiao' },
      { bio: '' },
      { websiteUrl: null },
      { email: '' },
      { public_repo_count: 4 },
      { followers_count: 1 },
      { followings_count: 1 },
      { createdAt: '2018-09-04T01:08:13Z' }
    ]}/>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center'
  },
  item: {
    backgroundColor: '#f9c2ff',
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16
  },
  title: {
    fontSize: 32
  }
})

export { HomeScreenWrapper, HomeScreen }
