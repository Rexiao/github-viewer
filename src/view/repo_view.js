import 'react-native-gesture-handler'
import React, { useEffect, useState } from 'react'
import { FlatList, Button, Text, View, ActivityIndicator } from 'react-native'
import token from './../models/token'
import { Repo, queryText } from './../models/repo'
import { styles } from './style'

/**
 * item function used in renderItem
 * @param {Object} props l:key
 *                       r:value
 *                       nav: react navigation
 */
const Item = function itemFunction (props) {
  return (
    <View style={styles.item}>
      <Text>repo_name: {props.data.repo_name}</Text>
      <Text>owner_name: {props.data.owner_name}</Text>
      <Text>description: {props.data.description}</Text>
    </View>
  )
}

/**
 * wrapper used in flat list
 * @param {Object} props l:key
 *                       r:value
 *                       nav: react navigation
 */
const renderItem = ({ item }) => (
  <Item data={item} />
)

/**
 * function for repo screen
 * @param {Object} props props.nav react navigation
 *                       props.needFetch if the view need actually fetch data from github
 *                       props.needLoadScreen if the view only show activity indicator
 *                       props.fakeData the fake data used in view if needFetch is false
 */
function RepoScreen (props) {
  const [isLoading, setLoading] = useState(true)
  const [data, setData] = useState([])
  const variables = { total_repo_count: 15 }
  useEffect(() => {
    if (props.needFetch) {
      Repo.fetchWrapper(token, queryText, variables, setData, setLoading)
    } else {
      // console.log('null becuase of fake!')
      if (props.needLoadScreen) {
        setLoading(true)
      } else if (!props.fakeData) {
        setLoading(false)
        setData(null)
      } else {
        setLoading(false)
        setData(props.fakeData)
      }
    }
  }, [])

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      {isLoading ? <ActivityIndicator/> : (
        !data ? (<Text>load data fails. If the reason is related to network, check console.error. Otherwise, it is becuase of invalid data format</Text>) : (
          <FlatList
            data={data}
            renderItem={renderItem}
            keyExtractor={(item, index) => index.toString()}
          />
        )
      )}
      <Button onPress={() => props.nav.goBack()} title="Go back profile" />
    </View>
  )
}

function RepoScreenWrapper ({ navigation }) {
  return (
    <RepoScreen needLoadScreen={false} nav={navigation} needFetch={true} fakeData={[
      { repo_name: 'learn', owner_name: 'Rexiao', description: null },
      { repo_name: 'fa18-hw-ref', owner_name: 'Rexiao', description: null },
      {
        repo_name: 'flappy_hotdog',
        owner_name: 'Rexiao',
        description: null
      },
      {
        repo_name: 'mvc-droids',
        owner_name: 'Rexiao',
        description: 'The mvc tutorial'
      }
    ]}/>
  )
}

export { RepoScreen, RepoScreenWrapper }
