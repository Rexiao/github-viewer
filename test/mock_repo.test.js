import 'react-native'
import React from 'react'
import { RepoScreen } from './../src/view/repo_view'
// test renderer must be required after react-native.
import { render, cleanup, waitFor } from '@testing-library/react-native'

beforeEach(() => {
  fetch.resetMocks()
})
afterEach(cleanup)

test('renders repo view fetch success', async () => {
  fetch.mockResponseOnce(JSON.stringify({
    data: {
      viewer: {
        repositories: {
          totalCount: 4,
          nodes: [
            {
              description: null,
              name: 'learn',
              owner: {
                login: 'Rexiao'
              }
            },
            {
              description: null,
              name: 'fa18-hw-ref',
              owner: {
                login: 'Rexiao'
              }
            },
            {
              description: null,
              name: 'flappy_hotdog',
              owner: {
                login: 'Rexiao'
              }
            },
            {
              description: 'The mvc tutorial',
              name: 'mvc-droids',
              owner: {
                login: 'Rexiao'
              }
            }
          ]
        }
      }
    }
  }))
  const { toJSON } = render(<RepoScreen needFetch={true}/>)
  await waitFor(() => {
    expect(toJSON()).toMatchSnapshot()
  })
})

test('renders repo view fetch fail becuase promise is rejected', async () => {
  fetch.mockReject(() => Promise.reject('API is down'))
  const { toJSON } = render(<RepoScreen needFetch={true}/>)
  await waitFor(() => {
    expect(toJSON()).toMatchSnapshot()
  })
})

test('renders repo view fetch fail because data is not in correct format', async () => {
  fetch.mockResponseOnce(JSON.stringify({
    viewer: {
      repositories: {
        totalCount: 4,
        nodes: [
          {
            description: null,
            name: 'learn',
            owner: {
              login: 'Rexiao'
            }
          },
          {
            description: null,
            name: 'fa18-hw-ref',
            owner: {
              login: 'Rexiao'
            }
          },
          {
            description: null,
            name: 'flappy_hotdog',
            owner: {
              login: 'Rexiao'
            }
          },
          {
            description: 'The mvc tutorial',
            name: 'mvc-droids',
            owner: {
              login: 'Rexiao'
            }
          }
        ]
      }
    }
  }))
  const { toJSON } = render(<RepoScreen needFetch={true}/>)
  await waitFor(() => {
    expect(toJSON()).toMatchSnapshot()
  })
})

test('renders repo view fetch fail because data is not in correct format(repositiries)', async () => {
  fetch.mockResponseOnce(JSON.stringify({
    data: {
      viewer: {
        totalCount: 4,
        nodes: [
          {
            description: null,
            name: 'learn',
            owner: {
              login: 'Rexiao'
            }
          },
          {
            description: null,
            name: 'fa18-hw-ref',
            owner: {
              login: 'Rexiao'
            }
          },
          {
            description: null,
            name: 'flappy_hotdog',
            owner: {
              login: 'Rexiao'
            }
          },
          {
            description: 'The mvc tutorial',
            name: 'mvc-droids',
            owner: {
              login: 'Rexiao'
            }
          }
        ]
      }
    }
  }))
  const { toJSON } = render(<RepoScreen needFetch={true}/>)
  await waitFor(() => {
    expect(toJSON()).toMatchSnapshot()
  })
})

test('renders mian view fetch success but all data is empty', async () => {
  fetch.mockResponseOnce(JSON.stringify({
    data: {
      viewer: {
        repositories: {
          totalCount: 4,
          nodes: []
        }
      }
    }
  }))
  const { toJSON } = render(<RepoScreen needFetch={true}/>)
  await waitFor(() => {
    expect(toJSON()).toMatchSnapshot()
  })
})
