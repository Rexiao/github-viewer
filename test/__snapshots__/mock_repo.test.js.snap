// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders mian view fetch success but all data is empty 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go back profile
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders repo view fetch fail because data is not in correct format 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go back profile
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders repo view fetch fail because data is not in correct format(repositiries) 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go back profile
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders repo view fetch fail becuase promise is rejected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go back profile
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders repo view fetch success 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "description": null,
          "owner_name": "Rexiao",
          "repo_name": "learn",
        },
        Object {
          "description": null,
          "owner_name": "Rexiao",
          "repo_name": "fa18-hw-ref",
        },
        Object {
          "description": null,
          "owner_name": "Rexiao",
          "repo_name": "flappy_hotdog",
        },
        Object {
          "description": "The mvc tutorial",
          "owner_name": "Rexiao",
          "repo_name": "mvc-droids",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#f9c2ff",
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text>
            repo_name: 
            learn
          </Text>
          <Text>
            owner_name: 
            Rexiao
          </Text>
          <Text>
            description: 
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#f9c2ff",
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text>
            repo_name: 
            fa18-hw-ref
          </Text>
          <Text>
            owner_name: 
            Rexiao
          </Text>
          <Text>
            description: 
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#f9c2ff",
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text>
            repo_name: 
            flappy_hotdog
          </Text>
          <Text>
            owner_name: 
            Rexiao
          </Text>
          <Text>
            description: 
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#f9c2ff",
              "marginHorizontal": 16,
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text>
            repo_name: 
            mvc-droids
          </Text>
          <Text>
            owner_name: 
            Rexiao
          </Text>
          <Text>
            description: 
            The mvc tutorial
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go back profile
      </Text>
    </View>
  </View>
</View>
`;
